### ISSUE TITLE
`instanceof Array` in `hyperscript` can fail

### ISSUE THREAD

**Issue by arienkock** [lhorie/mithril.js#1390]
version: rewrite branch

### Description:

While running mithril in Nashorn (JS engine on JVM) got an error using the hyperscript helper function. It is fixed by changing the line which inspects the arguments to use `Array.isArray` instead of `instaneof`:

```javascript
	var attrs, children, childrenIndex
	if (arguments[1] == null || typeof arguments[1] === "object" && arguments[1].tag === undefined && !(Array.isArray(arguments[1]))) {
		attrs = arguments[1]
		childrenIndex = 2
	}
```

Apparently this can occur in browsers too:
http://stackoverflow.com/questions/28779255/is-instanceof-array-better-than-isarray-in-javascript

### Steps to Reproduce:

Difficult since it's in my custom application running in Nashorn JS engine. Not useful to reproduce, but here is 'proof' (see the value of eval variable):

![image](https://cloud.githubusercontent.com/assets/2857600/20031228/3ee8f97a-a374-11e6-9c70-6ba048f08906.png)

**Comment by valtron** [comment]
I've also ran into this, because of data that was created in a different frame.

**Comment by isiahmeadows** [comment]
@lhorie Should we just scour both code bases and replace all `instanceof Array`s with `Array.isArray`? The latter also have the advantages of being faster (even in V8 despite not having code generation) and working cross-realm.

**Comment by isiahmeadows** [comment]
@valtron Note that Nashorn is technically unsupported. Not that I'm adverse to fixing this, I'm just giving you a heads up on it.

**Comment by valtron** [comment]
@isiahmeadows Not in Nashorn, just a regular browser. (Worked around it with `JSON.parse(JSON.stringify(...))`, but took a while to figure out what was happening: `m('div', data)` errored, `m('div', data.slice(0))` errored, but `m('div', [data[0], data[1], ...])` worked 😕.)

**Comment by isiahmeadows** [comment]
@valtron I meant @arienkock. Sorry about that. :-\

**Comment by pygy** [comment]
The use of `typeof` and `instanceof` is [deliberate for perf reasons](https://github.com/lhorie/mithril.js/blob/634c0949fe67501d9a9c00a70d9d9a4fbc6e4554/CONTRIBUTING.md#why-does-the-mithril-codebase-use-a-mix-of-instanceof-and-typeof-checks-instead-of-objectprototypetostringcall-arrayisarray-etc-would-a-pr-to-refactor-those-checks-be-welcome), that being said, I was also puzzled by the cross-realm regression (since v0.x uses `({}).toString.call` for type checking).

Another thing going for `Array.isArray` is that it is [supported by IE9](http://kangax.github.io/compat-table/es5).

**Comment by isiahmeadows** [comment]
@pygy 

> Another thing going for `Array.isArray` is that it is supported by IE9.

Which make it 100% safe to replace all instances of `instanceof Array`. (All that don't support it are practically obsolete at this point, and it's trivial to polyfill.)

**Comment by barneycarroll** [comment]
> Another thing going for `Array.isArray` is that it is [supported by IE9](http://kangax.github.io/compat-table/es5/).

@pygy @isiahmeadows it seems you 2 are in implicit agreement that IE<9 is no longer supported. @lhorie seems [ambivalent on this issue](https://github.com/lhorie/mithril.js/issues/1418#issuecomment-262794456). If we're OK with that, #1423 does the job. If not, the polyfill (conditionally forking to `Object.prototype.toString` checks as before) is trivial to add back in.

**Comment by pygy** [comment]
@barneycarroll see my [reply in #1423](https://github.com/lhorie/mithril.js/issues/1418#issuecomment-262828742)

**Comment by barneycarroll** [comment]
Fixed in f149003 via #1423
